### DISCUSSION TITLE
An issue of sqlite3_key function

### DISCUSSION THREAD

**Discussion by pal98** [utelle/SQLite3MultipleCiphers#103] [Q&A] [answered]
Hello:
     I am using "sqlite3_key" and "sqlite3_rekey" for my project, it nearly works, only one issue.
     Here's my code :

```C++
    /* function of open a datebase with password */
    bool SQLite3DB::OpenDb(const char *dbName, const char *pw)
    {
	    int nResult = sqlite3_open(dbName, &db);
	    nResult = sqlite3_key(db, pw, strlen(pw));
            .......
     }
     
     /* function of reset password for a datebase*/
    bool SQLite3DB::SetDbPassword(const char *pw)
    {
	    if (sqlite3_rekey(db, pw, strlen(pw)) == SQLITE_OK)
		    return true;
	    else
		    return false;
    }
```
The problem is, when I call the function of "sqlite3_key", it always returns 0 (nResult  is always 0), no matter the password (pw value) is correct or not. 

While the function of "sqlite3_rekey" returns the value of 0 if the password is correct(for "sqlite3_key" function) and returns Nonzero value if the password is wrong.

Please give me some suggestion on how to determine whether the datebase password is correct or not by call "sqlite3_key" function.

**Comment by utelle** [comment, marked as answer]
> The problem is, when I call the function of "sqlite3_key", it always returns 0 (nResult is always 0), no matter the password (pw value) is correct or not.

Well, this is by design. The function `sqlite3_key` does not check whether the database is actually encrypted or not. This is documented in the description of the _official_ [**S**QLite **E**ncryption **E**xtension (SEE)](https://www.sqlite.org/see/doc/release/www/readme.wiki) - and this project does not change the behaviour. Here is the relevant quote from the documentation:

> > If you specify an incorrect key, you will not get an error message right away. But the first time you try to access the database you will get an `SQLITE_NOTADB` error with a message of "file is encrypted or is not a database". 

If you want to check whether the given password is correct or not, you have to issue an SQL command that actually accesses the database (like for example `SELECT count(*) FROM SQLITE_SCHEMA;`. If you get the `SQLITE_NOTADB` error, you can then display a message to the user, ask for the correct password, and repeat the call to `sqlite3_key`.

> While the function of "sqlite3_rekey" returns the value of 0 if the password is correct(for "sqlite3_key" function) and returns Nonzero value if the password is wrong.

This is not surprising, because `sqlite3_rekey` **accesses** the database file, and if the password given in `sqlite3_key` was wrong, you will get an error, of course.

> Please give me some suggestion on how to determine whether the datebase password is correct or not by call "sqlite3_key" function.

As said `sqlite3_key` does not check whether the password is correct or not. `sqlite3_key` will only fail, if the encryption scheme could not be set up correctly. For example, you can't set up encryption for a memory database or a temporary database.
